arwen/host/asyncCall.go

```diff
@@ -264,3 +265,388 @@ func (host *vmHost) computeDataLengthFromArguments(function string, arguments []
+	storage := host.Storage()
+
+	storageKey := arwen.CustomStorageKey(arwen.AsyncDataPrefix, runtime.GetOriginalTxHash())
+	buff := storage.GetStorage(storageKey)
+
+	asyncInfo := &arwen.AsyncContextInfo{}
+	err := json.Unmarshal(buff, &asyncInfo)
+	if err != nil {
+		return err
+	}
+
+	vmInput := runtime.GetVMInput()
+	var asyncCallPosition int
+	var currentContextIdentifier string
+	for contextIdentifier, asyncContext := range asyncInfo.AsyncContextMap {
+		for position, asyncCall := range asyncContext.AsyncCalls {
+			if bytes.Equal(vmInput.CallerAddr, asyncCall.Destination) {
+				asyncCallPosition = position
+				currentContextIdentifier = contextIdentifier
+				break
+			}
+		}
+
+		if len(currentContextIdentifier) > 0 {
+			break
+		}
+	}
+
+	if len(currentContextIdentifier) == 0 {
+		return arwen.ErrCallBackFuncNotExpected
+	}
+
+	// Remove current async call from the pending list
+	currentContextCalls := asyncInfo.AsyncContextMap[currentContextIdentifier].AsyncCalls
+	currentContextCalls[asyncCallPosition] = currentContextCalls[len(currentContextCalls)-1]
+	currentContextCalls[len(currentContextCalls)-1] = nil
+	currentContextCalls = currentContextCalls[:len(currentContextCalls)-1]
+
+	if len(currentContextCalls) == 0 {
+		// call OUR callback for resolving a full context
+		delete(asyncInfo.AsyncContextMap, currentContextIdentifier)
+	}
+
+	// If we are still waiting for callbacks we return
+	if len(asyncInfo.AsyncContextMap) > 0 {
+		return nil
+	}
+
+	_, err = storage.SetStorage(storageKey, nil)
+	if err != nil {
+		return err
+	}
+
+	// Now figure out if we can execute the callback here or different shard
+	if !host.canExecuteSynchronouslyOnDest(asyncInfo.AsyncInitiator.CallerAddr) {
+		err = host.sendStorageCallbackToDestination(asyncInfo.AsyncInitiator)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	// The caller is in the same shard, execute it's callback
+	callbackCallInput, err := host.createCallbackContractCallInput(
+		host.Output().GetVMOutput(),
+		asyncInfo.AsyncInitiator.CallerAddr,
+		arwen.CallbackDefault,
+		nil,
+	)
+	if err != nil {
+		return err
+	}
+
+	callbackVMOutput, _, callBackErr := host.ExecuteOnDestContext(callbackCallInput)
+	err = host.processCallbackVMOutput(callbackVMOutput, callBackErr)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+/**
+ * setupAsyncCallsGas sets the gasLimit for each async call with the amount of gas provided by the
+ *  SC developer. The remaining gas is split between the async calls where the developer
+ *  did not specify any gas amount
+ */
+func (host *vmHost) setupAsyncCallsGas(asyncInfo *arwen.AsyncContextInfo) error {
+	gasLeft := host.Metering().GasLeft()
+	gasNeeded := uint64(0)
+	callsWithZeroGas := uint64(0)
+
+	for identifier, asyncContext := range asyncInfo.AsyncContextMap {
+		for index, asyncCall := range asyncContext.AsyncCalls {
+			gasNeeded += asyncCall.ProvidedGas
+			if gasNeeded > gasLeft {
+				return arwen.ErrNotEnoughGas
+			}
```

Comment:
it's better as it is, the subtraction can yield < 0 causing underflow

Reply:
We could implement a safe math component like the solidity library if you'd think it would be useful

Reply:
Added a new math package and an overflow file. Currently it only contains `AddUint64` function because that's the one we need now, but on the same model we can implement new functions when we need them